In [2]:
## System requirements
import pandas as pd
import numpy as np
import json
import urllib.request
import requests

In [3]:
df = pd.read_csv('covid-19-data/public/data/owid-covid-data.csv')

In [4]:
df.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2019-12-31          0.0        0.0   
1      AFG      Asia  Afghanistan  2020-01-01          0.0        0.0   
2      AFG      Asia  Afghanistan  2020-01-02          0.0        0.0   
3      AFG      Asia  Afghanistan  2020-01-03          0.0        0.0   
4      AFG      Asia  Afghanistan  2020-01-04          0.0        0.0   

   total_deaths  new_deaths  total_cases_per_million  new_cases_per_million  \
0           0.0         0.0                      0.0                    0.0   
1           0.0         0.0                      0.0                    0.0   
2           0.0         0.0                      0.0                    0.0   
3           0.0         0.0                      0.0                    0.0   
4           0.0         0.0                      0.0                    0.0   

   ...  aged_70_older  gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
0  ...          1.337        1803.987              NaN                597.029   
1  ...          1.337        1803.987              NaN                597.029   
2  ...          1.337        1803.987              NaN                597.029   
3  ...          1.337        1803.987              NaN                597.029   
4  ...          1.337        1803.987              NaN                597.029   

   diabetes_prevalence  female_smokers  male_smokers  handwashing_facilities  \
0                 9.59             NaN           NaN                  37.746   
1                 9.59             NaN           NaN                  37.746   
2                 9.59             NaN           NaN                  37.746   
3                 9.59             NaN           NaN                  37.746   
4                 9.59             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  
0                         0.5            64.83  
1                         0.5            64.83  
2                         0.5            64.83  
3                         0.5            64.83  
4                         0.5            64.83  

[5 rows x 36 columns]

In [5]:
df.shape

(36984, 36)

In [6]:
df.date.nunique()

229

In [7]:
df.location.nunique()

212

In [8]:
df.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_cases_per_million',
       'new_cases_per_million', 'total_deaths_per_million',
       'new_deaths_per_million', 'new_tests', 'total_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'tests_per_case', 'positive_rate', 'tests_units', 'stringency_index',
       'population', 'population_density', 'median_age', 'aged_65_older',
       'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cardiovasc_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand',
       'life_expectancy'],
      dtype='object')

#####  Data Information:
We have daily data for 212 locations.
Variables: 
['iso_code', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_cases_per_million',
       'new_cases_per_million', 'total_deaths_per_million',
       'new_deaths_per_million', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'tests_units',
       'stringency_index', 'population', 'population_density', 'median_age',
       'aged_65_older', 'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cvd_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_100k']

In [9]:
df_latest = df.groupby('location').agg({'date': 'max', 'total_cases_per_million': 'max', 'new_cases_per_million': 'max', 'total_deaths_per_million': 'max', 'total_tests_per_thousand': 'max', 'population_density': 'max', 'total_cases': 'max', 'total_deaths': 'max'}).reset_index()

In [11]:
df_latest.describe([0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]).round(2)

total_cases_per_million  new_cases_per_million  \
count                   210.00                 210.00   
mean                   3440.26                 208.10   
std                    5129.39                 450.05   
min                       2.75                   0.42   
10%                      86.63                   5.66   
25%                     303.14                  19.99   
50%                    1297.73                  66.78   
75%                    4580.75                 226.07   
90%                    8524.37                 506.71   
95%                   14429.93                 852.40   
99%                   21077.98                1759.77   
max                   39753.42                4944.38   

       total_deaths_per_million  total_tests_per_thousand  population_density  \
count                    210.00                     89.00              200.00   
mean                      94.55                     92.41              360.37   
std                      173.75                    123.75             1573.68   
min                        0.00                      1.66                0.14   
10%                        0.08                      5.73               15.31   
25%                        4.58                     13.57               37.62   
50%                       22.51                     55.90               87.25   
75%                       97.68                    117.38              213.88   
90%                      282.50                    202.23              438.66   
95%                      485.25                    281.12              799.75   
99%                      775.53                    600.26             7048.47   
max                     1237.55                    727.42            19347.50   

       total_cases  total_deaths  
count       211.00        211.00  
mean     201077.28       7233.77  
std     1527542.38      54338.14  
min           3.00          0.00  
10%         133.00          1.00  
25%         824.50         11.50  
50%        4988.00         93.00  
75%       41336.00        665.00  
90%      168290.00       6021.00  
95%      340819.00      17414.50  
99%     3200587.20     101461.50  
max    21213649.00     760421.00

In [12]:
## Plotting
from bokeh.plotting import figure, output_file, show
from bokeh.models import ColumnDataSource
from bokeh.models.tools import HoverTool

In [13]:
source = ColumnDataSource(df_latest)
p = figure()
p.circle(x='total_cases', y='population_density',
         source=source,
         size=10, color='green')
hover = HoverTool()
hover.tooltips=[
    ('total cases', '@total_cases'),
    ('population_density', '@population_density'),
    ('total deaths', '@total_deaths')
]

p.add_tools(hover)

show(p)

In [33]:
df_latest.sort_values(['total_cases_per_million'], ascending=[1])

location        date  total_cases_per_million  \
149  Papua New Guinea  2020-05-28                    0.894   
111           Lesotho  2020-05-28                    0.934   
4              Angola  2020-05-28                    2.160   
108              Laos  2020-05-28                    2.611   
206           Vietnam  2020-05-28                    3.359   
..                ...         ...                      ...   
204           Vatican  2020-05-28                14833.127   
156             Qatar  2020-05-28                16989.233   
163        San Marino  2020-05-28                19653.486   
87          Hong Kong  2020-05-26                      NaN   
92      International  2020-03-10                      NaN   

     new_cases_per_million  total_deaths_per_million  \
149                  0.559                     0.000   
111                  0.467                     0.000   
4                    0.274                     0.122   
108                  0.550                     0.000   
206                  0.267                     0.000   
..                     ...                       ...   
204               4944.376                     0.000   
156                635.183                    10.413   
163               1060.758                  1237.551   
87                     NaN                       NaN   
92                     NaN                       NaN   

     total_tests_per_thousand  population_density  total_cases  total_deaths  
149                       NaN              18.220            8             0  
111                       NaN              73.562            2             0  
4                         NaN              23.890           71             4  
108                       NaN              29.715           19             0  
206                     2.681             308.127          327             0  
..                        ...                 ...          ...           ...  
204                       NaN                 NaN           12             0  
156                    71.860             227.322        48947            30  
163                       NaN             556.667          667            42  
87                     29.618            7039.714            0             0  
92                        NaN                 NaN          705             7  

[212 rows x 9 columns]

#####  Indian Regression Model

In [35]:
df_ind = df[df.location=='India']

In [38]:
df_ind_1 = df_ind[['date', 'total_cases', 'new_cases', 'total_deaths', 'new_deaths', 'total_cases_per_million', 'new_cases_per_million',  'total_deaths_per_million',
       'new_deaths_per_million', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'tests_units']]

In [39]:
df_ind_1

date  total_cases  new_cases  total_deaths  new_deaths  \
8748  2019-12-31            0          0             0           0   
8749  2020-01-01            0          0             0           0   
8750  2020-01-02            0          0             0           0   
8751  2020-01-03            0          0             0           0   
8752  2020-01-04            0          0             0           0   
...          ...          ...        ...           ...         ...   
8892  2020-05-24       131868       6767          3867         147   
8893  2020-05-25       138845       6977          4021         154   
8894  2020-05-26       145380       6535          4167         146   
8895  2020-05-27       151767       6387          4337         170   
8896  2020-05-28       158333       6566          4531         194   

      total_cases_per_million  new_cases_per_million  \
8748                    0.000                  0.000   
8749                    0.000                  0.000   
8750                    0.000                  0.000   
8751                    0.000                  0.000   
8752                    0.000                  0.000   
...                       ...                    ...   
8892                   95.556                  4.904   
8893                  100.612                  5.056   
8894                  105.347                  4.735   
8895                  109.976                  4.628   
8896                  114.734                  4.758   

      total_deaths_per_million  new_deaths_per_million  total_tests  \
8748                     0.000                   0.000          NaN   
8749                     0.000                   0.000          NaN   
8750                     0.000                   0.000          NaN   
8751                     0.000                   0.000          NaN   
8752                     0.000                   0.000          NaN   
...                        ...                     ...          ...   
8892                     2.802                   0.107    2943421.0   
8893                     2.914                   0.112    3033591.0   
8894                     3.020                   0.106    3126119.0   
8895                     3.143                   0.123    3242160.0   
8896                     3.283                   0.141    3362136.0   

      new_tests  total_tests_per_thousand  new_tests_per_thousand  \
8748        NaN                       NaN                     NaN   
8749        NaN                       NaN                     NaN   
8750        NaN                       NaN                     NaN   
8751        NaN                       NaN                     NaN   
8752        NaN                       NaN                     NaN   
...         ...                       ...                     ...   
8892   108623.0                     2.133                   0.079   
8893    90170.0                     2.198                   0.065   
8894    92528.0                     2.265                   0.067   
8895   116041.0                     2.349                   0.084   
8896   119976.0                     2.436                   0.087   

      new_tests_smoothed  new_tests_smoothed_per_thousand     tests_units  
8748                 NaN                              NaN             NaN  
8749                 NaN                              NaN             NaN  
8750                 NaN                              NaN             NaN  
8751                 NaN                              NaN             NaN  
8752                 NaN                              NaN             NaN  
...                  ...                              ...             ...  
8892            102254.0                            0.074  samples tested  
8893            104400.0                            0.076  samples tested  
8894            103122.0                            0.075  samples tested  
8895            104253.0                            0.07

### Linear Regression on Corona Cases for India

In [15]:
df.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_cases_per_million',
       'new_cases_per_million', 'total_deaths_per_million',
       'new_deaths_per_million', 'new_tests', 'total_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'tests_per_case', 'positive_rate', 'tests_units', 'stringency_index',
       'population', 'population_density', 'median_age', 'aged_65_older',
       'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cardiovasc_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand',
       'life_expectancy'],
      dtype='object')

In [78]:
df_ind = df[df.iso_code == 'IND']

In [79]:
## Finding date with first case and filtering the data from that point
df_ind = df_ind[df_ind.total_cases > 0]

In [80]:
## first case reported on 30th January 2020
df_ind.location.unique()

array(['India'], dtype=object)

In [81]:
df_ind = df_ind[['date', 'total_cases', 'new_cases', 'total_deaths', 
                 'new_deaths',  'new_tests', 'total_tests', 'tests_per_case', 
                 'positive_rate', 'population', 'population_density', 
                'median_age', 'aged_65_older','stringency_index',
       'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cardiovasc_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand',
       'life_expectancy']]

In [50]:
df_ind.life_expectancy.unique()

array([450.419])

In [31]:
df_var = df[['location', 'date', 'total_cases', 'new_cases', 'total_deaths', 
                 'new_deaths',  'new_tests', 'total_tests', 'tests_per_case', 
                 'positive_rate', 'population', 'population_density', 
                'median_age', 'aged_65_older',
       'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cardiovasc_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand',
       'life_expectancy']]

In [32]:
df_var.head()

location        date  total_cases  new_cases  total_deaths  new_deaths  \
0  Afghanistan  2019-12-31          0.0        0.0           0.0         0.0   
1  Afghanistan  2020-01-01          0.0        0.0           0.0         0.0   
2  Afghanistan  2020-01-02          0.0        0.0           0.0         0.0   
3  Afghanistan  2020-01-03          0.0        0.0           0.0         0.0   
4  Afghanistan  2020-01-04          0.0        0.0           0.0         0.0   

   new_tests  total_tests  tests_per_case  positive_rate  ...  aged_70_older  \
0        NaN          NaN             NaN            NaN  ...          1.337   
1        NaN          NaN             NaN            NaN  ...          1.337   
2        NaN          NaN             NaN            NaN  ...          1.337   
3        NaN          NaN             NaN            NaN  ...          1.337   
4        NaN          NaN             NaN            NaN  ...          1.337   

   gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
0        1803.987              NaN                597.029   
1        1803.987              NaN                597.029   
2        1803.987              NaN                597.029   
3        1803.987              NaN                597.029   
4        1803.987              NaN                597.029   

   diabetes_prevalence  female_smokers  male_smokers  handwashing_facilities  \
0                 9.59             NaN           NaN                  37.746   
1                 9.59             NaN           NaN                  37.746   
2                 9.59             NaN           NaN                  37.746   
3                 9.59             NaN           NaN                  37.746   
4                 9.59             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  
0                         0.5            64.83  
1                         0.5            64.83  
2                         0.5            64.83  
3                         0.5            64.83  
4                         0.5            64.83  

[5 rows x 24 columns]

In [52]:
df_cp = df_var.dropna()

In [55]:
df_cp.location.unique()

array(['Bangladesh', 'Colombia', 'Costa Rica', 'El Salvador', 'Ethiopia',
       'Ghana', 'India', 'Indonesia', 'Kazakhstan', 'Kenya', 'Mexico',
       'Myanmar', 'Nepal', 'Pakistan', 'Paraguay', 'South Africa',
       'Thailand', 'Togo', 'Tunisia', 'Uganda', 'Vietnam', 'Zimbabwe'],
      dtype=object)

In [59]:
df_continent = df.groupby(['continent', 'date']).agg({'new_cases': 'sum', 'total_cases': 'sum', 
                                                      'total_deaths': 'sum', 'new_deaths': 'sum', 
                                                      'new_tests': 'sum', 
                                                      'total_tests': 'sum', 'positive_rate': 'mean', 
                                                     'stringency_index': 'mean', 'population_density': 'mean',
                                                     'median_age': 'mean', 'aged_70_older': 'mean', 
                                                      'gdp_per_capita': 'mean', 'extreme_poverty': 'mean',
                                                      'cardiovasc_death_rate': 'mean', 'diabetes_prevalence':'mean'}).reset_index()

In [67]:
df_continent[df_continent.date == '2020-08-15'].sort_values(by = 'new_cases', ascending = False)

continent        date  new_cases  total_cases  total_deaths  \
457            Asia  2020-08-15    95916.0    5512064.0      117229.0   
1373  South America  2020-08-15    84099.0    5136357.0      173028.0   
915   North America  2020-08-15    76850.0    6316513.0      242102.0   
686          Europe  2020-08-15    18686.0    2781889.0      173928.0   
228          Africa  2020-08-15    12689.0    1098277.0       25102.0   
1144        Oceania  2020-08-15      434.0      25040.0         408.0   

      new_deaths  new_tests  total_tests  positive_rate  stringency_index  \
457       1519.0        0.0          0.0            NaN               NaN   
1373      1935.0        0.0          0.0            NaN               NaN   
915       2067.0        0.0          0.0            NaN               NaN   
686        287.0        0.0          0.0            NaN             37.96   
228        422.0        0.0          0.0            NaN               NaN   
1144        14.0        0.0          0.0            NaN               NaN   

      population_density  median_age  aged_70_older  gdp_per_capita  \
457           445.049136   30.158696       3.942739    22185.363295   
1373           24.392000   30.216667       5.186000    13840.785250   
915           270.565545   33.128000       6.227960    21655.243704   
686           638.055255   41.982051      11.717763    33338.380756   
228           103.492453   21.210909       2.167527     5443.837132   
1144           75.732750   32.071429       5.978571    23315.180500   

      extreme_poverty  cardiovasc_death_rate  diabetes_prevalence  
457          5.767857             290.287261             9.351818  
1373         2.866667             187.323917             7.810833  
915          5.718182             197.066885            10.880968  
686          0.900000             233.428325             6.544186  
228         34.100000             287.297426             5.629444  
1144         0.950000             286.065333            16.114286

In [68]:
df.location.unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Anguilla', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin',
       'Bermuda', 'Bhutan', 'Bolivia', 'Bonaire Sint Eustatius and Saba',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'British Virgin Islands', 'Brunei', 'Bulgaria', 'Burkina Faso',
       'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde',
       'Cayman Islands', 'Central African Republic', 'Chad', 'Chile',
       'China', 'Colombia', 'Comoros', 'Congo', 'Costa Rica',
       "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curacao', 'Cyprus',
       'Czech Republic', 'Democratic Republic of Congo', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Ethiopia', 'Faeroe Islands', 'Falkland Isla

##### Visualizations for India

In [134]:
## Plotting
from bokeh.plotting import figure, output_file, show
from bokeh.models import ColumnDataSource
from bokeh.models.tools import HoverTool

In [75]:
## Filtering the data for suitable data matches and across variables
df_ind[df_ind.new_tests.isnull()].date.unique()

array(['2020-01-30', '2020-01-31', '2020-02-01', '2020-02-02',
       '2020-02-03', '2020-02-04', '2020-02-05', '2020-02-06',
       '2020-02-07', '2020-02-08', '2020-02-09', '2020-02-10',
       '2020-02-11', '2020-02-12', '2020-02-13', '2020-02-14',
       '2020-02-15', '2020-02-16', '2020-02-17', '2020-02-18',
       '2020-02-19', '2020-02-20', '2020-02-21', '2020-02-22',
       '2020-02-23', '2020-02-24', '2020-02-25', '2020-02-26',
       '2020-02-27', '2020-02-28', '2020-02-29', '2020-03-01',
       '2020-03-02', '2020-03-03', '2020-03-04', '2020-03-05',
       '2020-03-06', '2020-03-07', '2020-03-08', '2020-03-10',
       '2020-03-11', '2020-03-12', '2020-03-13', '2020-03-14',
       '2020-03-15', '2020-03-16', '2020-03-17', '2020-03-25',
       '2020-03-26', '2020-03-27', '2020-03-28', '2020-03-29',
       '2020-03-30', '2020-03-31', '2020-04-01', '2020-04-19',
       '2020-04-20', '2020-04-21', '2020-04-22', '2020-08-14',
       '2020-08-15'], dtype=object)

In [87]:
df_ind.columns

Index(['date', 'total_cases', 'new_cases', 'total_deaths', 'new_deaths',
       'new_tests', 'total_tests', 'tests_per_case', 'positive_rate',
       'population', 'population_density', 'median_age', 'aged_65_older',
       'stringency_index', 'aged_70_older', 'gdp_per_capita',
       'extreme_poverty', 'cardiovasc_death_rate', 'diabetes_prevalence',
       'female_smokers', 'male_smokers', 'handwashing_facilities',
       'hospital_beds_per_thousand', 'life_expectancy'],
      dtype='object')

In [86]:
df_ind['stringency_index'].fillna(79.63, inplace=True)

In [97]:
## For each variable give null dates
col_null = []
for col in df_ind.columns:
    if len(list(df_ind[df_ind[col].isnull()]['date'])) > 0:
        col_null.append(col)
        print (col, " null date values for dates: ", set(list(pd.to_datetime(df_ind[df_ind[col].isnull()]['date']).dt.month)))

new_tests  null date values for dates:  {1, 2, 3, 4, 8}
total_tests  null date values for dates:  {1, 2, 3, 4, 8}
tests_per_case  null date values for dates:  {8, 1, 2, 3}
positive_rate  null date values for dates:  {8, 1, 2, 3}


In [99]:
for col in df_ind.columns:
    if len(list(df_ind[df_ind[col].isnull()]['date'])) > 0:
        print (col, " null date values for dates: ", set(list(df_ind[df_ind[col].isnull()]['date'])))

new_tests  null date values for dates:  {'2020-02-07', '2020-02-20', '2020-03-27', '2020-03-01', '2020-03-06', '2020-02-22', '2020-03-08', '2020-03-16', '2020-02-08', '2020-03-07', '2020-03-04', '2020-04-01', '2020-02-06', '2020-03-03', '2020-08-15', '2020-03-11', '2020-03-05', '2020-03-02', '2020-02-09', '2020-02-21', '2020-02-27', '2020-02-15', '2020-02-13', '2020-03-14', '2020-01-30', '2020-04-22', '2020-02-17', '2020-03-26', '2020-08-14', '2020-03-10', '2020-02-18', '2020-02-01', '2020-02-05', '2020-02-11', '2020-02-29', '2020-02-14', '2020-02-25', '2020-02-28', '2020-03-13', '2020-03-28', '2020-03-12', '2020-03-15', '2020-02-26', '2020-03-31', '2020-02-03', '2020-02-16', '2020-01-31', '2020-02-23', '2020-02-19', '2020-03-30', '2020-04-21', '2020-02-12', '2020-02-02', '2020-03-25', '2020-02-10', '2020-03-29', '2020-04-20', '2020-04-19', '2020-02-24', '2020-03-17', '2020-02-04'}
total_tests  null date values for dates:  {'2020-02-07', '2020-02-20', '2020-03-27', '2020-03-01', '2020-

In [100]:
col_null

['new_tests', 'total_tests', 'tests_per_case', 'positive_rate']

In [101]:
df_impute = df_ind[['date','new_tests', 'total_tests', 'tests_per_case', 'positive_rate']]

In [103]:
df_impute['datetime'] = pd.to_datetime(df_ind['date'])
df_impute['month'] = df_impute['datetime'].dt.month

/Users/devanshsharma/anaconda3/envs/data_science/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/Users/devanshsharma/anaconda3/envs/data_science/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [110]:
df_imp_vals = df_impute.groupby('month')['new_tests', 'total_tests', 'tests_per_case', 'positive_rate'].mean().reset_index()

In [111]:
df_imp_vals

month      new_tests   total_tests  tests_per_case  positive_rate
0      1            NaN           NaN             NaN            NaN
1      2            NaN           NaN             NaN            NaN
2      3    1717.000000  1.744200e+04       39.078538       0.029692
3      4   30229.880000  3.591524e+05       25.854933       0.039100
4      5   94663.000000  2.288044e+06       22.962323       0.045903
5      6  166312.600000  6.169970e+06       13.705267       0.073567
6      7  339744.322581  1.343263e+07       10.073290       0.100000
7      8  641212.076923  2.347893e+07       10.465077       0.095923

In [107]:
df_impute['weekofyear'] = df_impute['datetime'].dt.week

/Users/devanshsharma/anaconda3/envs/data_science/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [108]:
df_impute.head()

date  new_tests  total_tests  tests_per_case  positive_rate  \
15527  2020-01-30        NaN          NaN             NaN            NaN   
15528  2020-01-31        NaN          NaN             NaN            NaN   
15529  2020-02-01        NaN          NaN             NaN            NaN   
15530  2020-02-02        NaN          NaN             NaN            NaN   
15531  2020-02-03        NaN          NaN             NaN            NaN   

        datetime  month  weekofyear  
15527 2020-01-30      1           5  
15528 2020-01-31      1           5  
15529 2020-02-01      2           5  
15530 2020-02-02      2           5  
15531 2020-02-03      2           6

In [112]:
df_imp_vals_week = df_impute.groupby('weekofyear')['new_tests', 'total_tests', 'tests_per_case', 'positive_rate'].mean().reset_index()

In [114]:
for col in col_null:
    df_imp_vals_week.rename(columns = {col: col+'_imp'}, inplace = True)

In [118]:
df_imp_vals_week
df_ind['weekofyear'] = pd.to_datetime(df_ind['date']).dt.week

In [120]:
df_ind_2 = pd.merge(df_ind, df_imp_vals_week, on=['weekofyear'], how = 'left')

In [123]:
df_ind_2.new_tests.fillna(df_ind_2.new_tests_imp, inplace= True)

In [126]:
df_ind_2.tests_per_case.fillna(df_ind_2.tests_per_case_imp, inplace= True)
df_ind_2.positive_rate.fillna(df_ind_2.positive_rate_imp, inplace= True)

In [127]:
df_ind_2.drop(columns=['total_tests'], inplace=True)

In [130]:
df_ind_2.shape

(198, 28)

In [132]:
df_ind_3 = df_ind_2.dropna()

In [133]:
df_ind_3.head()

date  total_cases  new_cases  total_deaths  new_deaths  new_tests  \
45  2020-03-16         93.0        3.0           2.0         0.0     1516.4   
46  2020-03-17        125.0       32.0           3.0         1.0     1516.4   
47  2020-03-18        137.0       12.0           3.0         0.0      191.0   
48  2020-03-19        165.0       28.0           3.0         0.0     1060.0   
49  2020-03-20        191.0       26.0           4.0         1.0     1325.0   

    tests_per_case  positive_rate    population  population_density  ...  \
45         62.7505          0.017  1.380004e+09             450.419  ...   
46         62.7505          0.017  1.380004e+09             450.419  ...   
47         62.7505          0.017  1.380004e+09             450.419  ...   
48         85.5980          0.012  1.380004e+09             450.419  ...   
49         67.8880          0.015  1.380004e+09             450.419  ...   

    female_smokers  male_smokers  handwashing_facilities  \
45             1.9          20.6                   59.55   
46             1.9          20.6                   59.55   
47             1.9          20.6                   59.55   
48             1.9          20.6                   59.55   
49             1.9          20.6                   59.55   

    hospital_beds_per_thousand  life_expectancy  weekofyear  new_tests_imp  \
45                        0.53            69.66          12         1516.4   
46                        0.53            69.66          12         1516.4   
47                        0.53            69.66          12         1516.4   
48                        0.53            69.66          12         1516.4   
49                        0.53            69.66          12         1516.4   

    total_tests_imp  tests_per_case_imp  positive_rate_imp  
45          15704.0             62.7505              0.017  
46          15704.0             62.7505              0.017  
47          15704.0             62.7505              0.017  
48          15704.0             62.7505              0.017  
49          15704.0             62.7505              0.017  

[5 rows x 28 columns]

##### Positive Rate Change across Weeks

In [149]:
weekly = df_ind_3.groupby('weekofyear')['positive_rate', 'new_cases', 'new_deaths', 'new_tests'].mean().reset_index()

In [142]:
from bokeh.plotting import figure, output_file, show
from bokeh.models import ColumnDataSource
from bokeh.palettes import Spectral3
output_file('simple_timeseries_plot.html')

#make sure MSNDATE is a datetime format
df_ind_3['date'] = pd.to_datetime(df_ind_3['date'])

grouped = df_ind_3.groupby('date')['new_cases', 'new_deaths'].sum()

source = ColumnDataSource(grouped)

p = figure(x_axis_type='datetime')

# p.line(x='date', y='new_tests', line_width=2, source=source, legend='new tests')
p.line(x='date', y='new_cases', line_width=2, source=source, color=Spectral3[1], legend='New Cases')
p.line(x='date', y='new_deaths', line_width=2, source=source, color=Spectral3[2], legend='New Deaths')

p.yaxis.axis_label = 'new cases and deaths with positive rate'

show(p)

/Users/devanshsharma/anaconda3/envs/data_science/lib/python3.6/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [146]:
# source = ColumnDataSource(weekly)
import pandas_bokeh
weekly.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['positive_rate'],
    xlabel='week',
    ylabel='positive rate',
    title='WoW change in Positive Rate'
)

Figure(id='2582', ...)

In [148]:
weekly.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['new_cases'],
    xlabel='week',
    ylabel='new cases',
    title='WoW change in New Cases'
)

Figure(id='2721', ...)

In [150]:
weekly.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['new_tests'],
    xlabel='week',
    ylabel='new tests',
    title='WoW change in New Test'
)

Figure(id='2860', ...)

In [153]:
df_us = df[df.iso_code == 'USA']

In [154]:
df_us['date'] = pd.to_datetime(df_us['date'])

/Users/devanshsharma/anaconda3/envs/data_science/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [155]:
df_us['weekofyear'] = df_us.date.dt.week

/Users/devanshsharma/anaconda3/envs/data_science/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [156]:
weekly_us = df_us.groupby(['weekofyear'])['positive_rate', 'new_cases', 'new_deaths', 'new_tests'].mean().reset_index()

In [157]:
weekly_us.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['positive_rate'],
    xlabel='week',
    ylabel='positive rate',
    title='WoW change in Positive Rate'
)

Figure(id='2999', ...)

In [158]:
weekly_us.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['new_cases'],
    xlabel='week',
    ylabel='new cases',
    title='WoW change in New Cases'
)

Figure(id='3138', ...)

In [159]:
weekly_us.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['new_tests'],
    xlabel='week',
    ylabel='new tests',
    title='WoW change in New Test'
)

Figure(id='3277', ...)

In [160]:
weekly_us.plot_bokeh(
    kind='line',
    x='weekofyear',
    y=['new_deaths'],
    xlabel='week',
    ylabel='new deaths',
    title='WoW change in New Deaths'
)

Figure(id='3416', ...)